### PR TITLE
#165939712 Connect users to PostgreSQL database

### PR DESCRIPTION
**Pull request by chidimo**
#### What does this PR do?

Setup database and connect users model

#### Description of Task to be completed?

1. Create a model class from which all data models inherit (simplifies DB operations)
1. Create a global settings file to abstract the definition of DB connection parameters
1. Use `node-postgres` to establish a DB connection and to run SQL queries

#### How should this be manually tested?

1. Clone the repo
1. `cd` into the root folder on a `cmd` shell and run `yarn install`. PowerShell is fine also.
1. Setup a PostgreSQL DB.
1. Define a `.env` file in the root of the project and supply values for the following parameters
    ```.env
    JWT_SECRET=<json-web-token-secret>
    PGPASSWORD=<postgres-db-password>
    PGHOST=<postgres-db-host>
    PGUSER=<postgres-db-user>
    PGPORT=<postgres-db-port>
    ```
1. Create a database with the name `testdb`.
1. Now run command `yarn test`

#### Any background context you want to provide?

Nil

#### What are the relevant pivotal tracker stories?

[165939712](https://www.pivotaltracker.com/story/show/165939712)

#### Screenshots (if appropriate)

Nil

#### Questions:

Nil

### CODE REVIEW

File: controllers/AuthController.js [hunk 1]
```diff
@@ -1,46 +1,95 @@
...
-                    },
-                    status: 'unverified',
-                    isAdmin: false,
-                    token: req.token,
+        const { email, password } = req.body;
+        const hashedPassword = bcrypt.hashSync(password, 8);
+        dev_logger(`token: ${req.token}`);
+
+        // check if user exists
+        const check_user_existence = async (email) => {
```
Comment (damafeez) [inline controllers/AuthController.js]: These are helper functions. You can move them into a `helper` folder or something. Let the controller handle the logic and not function declarations.

File: controllers/AuthController.js [hunk 2]
```diff
@@ -1,46 +1,95 @@
...
+        // add user to db
+        const create_user = async (email, password) => {
+            try {
+                await users_model.insert('(email, password)',
+                    `'${email}', '${password}'`);                    
+            }
+            catch (e) { return InternalServerError(req, res, e);}
+        };
+
+        const get_new_user = async (email) => {
```
Comment (damafeez) [inline controllers/AuthController.js]: These are helper functions. You can move them into a helper folder or something. Let the controller handle the logic and not function declarations.

File: controllers/AuthController.js [hunk 3]
```diff
@@ -1,46 +1,95 @@
...
+import Model from '../models/Model';
+import { InternalServerError, CriticalError } from '../utils/errorHandlers';
+import { dev_logger } from '../utils/loggers';
+
+// import { dev_logger, test_logger } from '../utils/loggers';
+
+const users_model = new Model('users');
 
 const AuthController = {
     signup: (req, res) => {
```
Comment (damafeez) [inline controllers/AuthController.js]: Make this function async so you can do await within its body

File: controllers/AuthController.js (hunk 1 above)
Comment (chidimo) [reply controllers/AuthController.js]: Done

File: controllers/AuthController.js [hunk 4]
```diff
@@ -1,46 +1,95 @@
...
+                const { rows } = await users_model.select(
+                    'id, email', `email='${email}'`);
+                const [ user, ] = rows;
+                if (user) return user;
+            }
+            catch (e) { return InternalServerError(req, res, e);}
+        };
+
+        // add user to db
+        const create_user = async (email, password) => {
```
Comment (damafeez) [inline controllers/AuthController.js]: These are helper functions. You can move them into a helper folder or something. Let the controller handle the logic and not function declarations.

File: controllers/AuthController.js [hunk 5]
```diff
@@ -1,46 +1,95 @@
...
-            });
-        }
+                return res
+                    .status(201)
+                    .json({ data: { ...rows[0], token: req.token } });
+            }
+            catch (e) { return InternalServerError(req, res, e);}
+        };
+
+        (async () => {
```
Comment (damafeez) [inline controllers/AuthController.js]: Make the outer function `async` you won't have a need for this IIFE

File: controllers/AuthController.js [hunk 6]
```diff
@@ -1,46 +1,95 @@
...
+                return res
+                    .status(404)
+                    .json({ error: `User with email ${email} already exists` });
+            }
+            await create_user(email, hashedPassword);
+            return get_new_user(email);
+        })();
     },
     
     signin: (req, res) => {
```
Comment (damafeez) [inline controllers/AuthController.js]: Make this function `async` so you can do `await` within its body

File: controllers/AuthController.js [hunk 7]
```diff
@@ -1,46 +1,95 @@
...
-            res.render( 'authentication', { title: 'Sign In' });
-        }
-        else {
-            const { email } = req.body;
-            const user = users.find(user => user.email === email);
-            const data = { ... user, token: req.token };
-            res.status(200).json({ data });
-        }
+        const { email } = req.body;
+        (async () => {
```
Comment (damafeez) [inline controllers/AuthController.js]: Make the outer function async you won't have a need for this IIFE

File: models/Model.js [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+    database: Settings.dbSettings().databaseName,
+    password: Settings.dbSettings().dbPassword
+};
+
+test_logger(`test db config ${JSON.stringify(config)}`);
+
+class Model {
+    constructor(table) {
+        this.table = table;
+        this.pool = new Pool(config);
```
Comment (damafeez) [inline models/Model.js:22]: You're aware this creates a new `Pool` for every model you create?
Is there a reason you're doing this and not using one `Pool` throughout the application?

File: controllers/AuthController.js [hunk 9]
```diff
@@ -1,46 +1,95 @@
...
+                const [ user, ] = rows;
+                if (user) return user;
+            }
+            catch (e) { return InternalServerError(req, res, e);}
+        };
+
+        // add user to db
+        const create_user = async (email, password) => {
+            try {
+                await users_model.insert('(email, password)',
```
Comment (damafeez) [inline controllers/AuthController.js]: You should be saving more than just the users `email` and `password` to the db.
Or, are these the only info you're expecting the client to send to the backend?

File: routes/index.js [hunk 10]
```diff
@@ -3,35 +3,34 @@ import express from 'express';
...
+import ParamterValidators from '../middleware/validators.js';
 
 const router = express.Router();
 
-/* GET home page. */
-router.get('/', AppController.index);
-router.get('/about', AppController.about);
-
 router.post('/auth/signup',
     AuthenticationMiddleware.generateToken,
```
Comment (damafeez) [inline routes/index.js]: You should not generate token before data validation. You can move this middleware to the bottom of the chain

**Review by damafeez — CHANGES_REQUESTED** [review]
You've done an awesome job @chidimo. I left you feedback on some actionable tasks, however.

File: controllers/UsersController.js [hunk 11]
```diff
@@ -1,92 +1,92 @@
...
-// const logger = message => (debug('dev')(message));
+const users_model = new Model('users');
 
 const UsersController = {
 
-    dashboard: (req, res) => {
-        res.render( 'dashboard', { title: 'Dashboard' });
+    get_user: (req, res) => {
+        const { id } = req.params;
+        (async () => {
```
Comment (damafeez) [inline controllers/UsersController.js]: Make the outer function async you won't have a need for this IIFE

File: controllers/UsersController.js [hunk 12]
```diff
@@ -1,92 +1,92 @@
...
-        
-        if (user === undefined) {
-            res.status(404).json({ error: 'User not found' });
-        }
-        else {
-            user.status = 'verified';
-            res.status(200).json({ data: user });
-        }
+        const { id } = req.params;
+        (async () => {
```
Comment (damafeez) [inline controllers/UsersController.js]: Make the outer function async you won't have a need for this IIFE

File: controllers/UsersController.js [hunk 13]
```diff
@@ -1,92 +1,92 @@
...
-        if (status) {
-            const data = users.filter(user => (user.status === status));
-            res.status(200).json({ data });
-        }
-        else {
-            const data = [];
-            users.map(user => data.push(user));
-            res.status(200).json({ data });
-        }
+        (async () => {
```
Comment (damafeez) [inline controllers/UsersController.js]: Make the outer function async you won't have a need for this IIFE

File: controllers/UsersController.js [hunk 14]
```diff
@@ -1,92 +1,92 @@
...
-        sanitizeBody('lastName').trim().escape(),
-        sanitizeBody('phone').trim().escape(),
-        sanitizeBody('home').trim().escape(),
-        sanitizeBody('office').trim().escape(),
-
-        (req, res) => {
-            const errors = validationResult(req);
-            if (!errors.isEmpty()) {
-                return res.status(422).json({ errors: errors.array() });
+        (async () => {
```
Comment (damafeez) [inline controllers/UsersController.js]: Make the outer function async you won't have a need for this IIFE

File: controllers/AuthController.js (hunk 9 above)
Comment (chidimo) [reply controllers/AuthController.js]: Yes. For registration purposes, as this simplifies the registration process.

Other information will be saved when they edit their profile.

I have an `update_user` function for that.

File: routes/index.js (hunk 10 above)
Comment (chidimo) [reply routes/index.js]: Data validation before token generation. Noted

File: models/Model.js (hunk 8 above)
Comment (chidimo) [reply models/Model.js:22]: I wasn't aware. But I'll look into it.

File: controllers/UsersController.js [hunk 15]
```diff
@@ -1,92 +1,92 @@
...
-import users from '../utils/sample.users';
-
-// const logger = message => (debug('dev')(message));
+const users_model = new Model('users');
 
 const UsersController = {
 
-    dashboard: (req, res) => {
-        res.render( 'dashboard', { title: 'Dashboard' });
+    get_user: (req, res) => {
```
Comment (damafeez) [inline controllers/UsersController.js]: Make this function async so you can do await within its body

File: controllers/UsersController.js [hunk 16]
```diff
@@ -1,92 +1,92 @@
...
+                return res.status(200).json({ data: rows[0] });
+            }
+            catch (e) {
+                return InternalServerError(req, res, e);
+            }
+        })();
     },
-    
+
     verify_user: (req, res) => {
```
Comment (damafeez) [inline controllers/UsersController.js]: Make this function async so you can do await within its body

File: controllers/UsersController.js [hunk 17]
```diff
@@ -1,92 +1,92 @@
...
+                    `id=${id}`
+                );
+                UsersController.get_user(req, res);
+            }
+            catch (e) { return InternalServerError(req, res, e); }
+        })();
     },
-    
+
     get_users: (req, res) => {
```
Comment (damafeez) [inline controllers/UsersController.js]: Make this function async so you can do await within its body

File: controllers/UsersController.js [hunk 18]
```diff
@@ -1,92 +1,92 @@
...
+                    );
+                }
+                return res.status(200).json({ data: data.rows });
+            }
+            catch (e) { return InternalServerError(req, res, e); }
+        })();
     },
 
-    get_user: (req, res) => {
+    update_user: (req, res) => {
```
Comment (damafeez) [inline controllers/UsersController.js]: Make this function async so you can do await within its body

File: models/Model.js (hunk 8 above)
Comment (chidimo) [reply models/Model.js:22]: I'll research the pros and cons of one vs multiple pools and see how I can implement the better.

**Review by damafeez — CHANGES_REQUESTED** [review]

File: models/Model.js (hunk 8 above)
Comment (damafeez) [reply models/Model.js:22]: That's ok @chidimo

File: controllers/AuthController.js (hunk 9 above)
Comment (damafeez) [reply controllers/AuthController.js]: We should not just save `email` and `password` on signup. Users are more comfortable with typical processes.

You can add, maybe, two feeds so users can find the flow a little different from the login flow

File: controllers/AuthController.js (hunk 2 above)
Comment (chidimo) [reply controllers/AuthController.js]: Done

File: controllers/AuthController.js (hunk 3 above)
Comment (chidimo) [reply controllers/AuthController.js]: Done

File: controllers/AuthController.js (hunk 5 above)
Comment (chidimo) [reply controllers/AuthController.js]: Done

File: controllers/AuthController.js (hunk 4 above)
Comment (chidimo) [reply controllers/AuthController.js]: Done

File: controllers/AuthController.js (hunk 6 above)
Comment (chidimo) [reply controllers/AuthController.js]: Done

File: controllers/AuthController.js (hunk 7 above)
Comment (chidimo) [reply controllers/AuthController.js]: Done

File: controllers/AuthController.js (hunk 9 above)
Comment (chidimo) [reply controllers/AuthController.js]: I'll look into this.

File: controllers/UsersController.js (hunk 11 above)
Comment (chidimo) [reply controllers/UsersController.js]: Done.

File: controllers/UsersController.js (hunk 12 above)
Comment (chidimo) [reply controllers/UsersController.js]: Done

File: controllers/UsersController.js (hunk 13 above)
Comment (chidimo) [reply controllers/UsersController.js]: Done

File: controllers/UsersController.js (hunk 14 above)
Comment (chidimo) [reply controllers/UsersController.js]: Done

File: controllers/UsersController.js (hunk 15 above)
Comment (chidimo) [reply controllers/UsersController.js]: Done

File: controllers/UsersController.js (hunk 16 above)
Comment (chidimo) [reply controllers/UsersController.js]: Done

File: controllers/UsersController.js (hunk 18 above)
Comment (chidimo) [reply controllers/UsersController.js]: Done

**Review by damafeez — APPROVED** [review]
This looks good to me @chidimo.
You can merge in.

**Pull request merged**